core/src/main/java/com/graphhopper/reader/osm/OSMNodeData.java

```diff
@@ -68,11 +65,12 @@ class OSMNodeData {
     private final PointAccess towerNodes;
 
     // this map stores an index for each OSM node we keep the node tags of. a value of -1 means there is no entry
-    // yet and a value of -2 means there was an entry but it was removed again
+    // yet.
     private final LongIntMap nodeTagIndicesByOsmNodeIds;
 
     // stores node tags
     private final List<Map<String, Object>> nodeTags;
+    private final LongSet nodesToBeSplit;
```

Comment:
In my last commit I added this LongSet to replace the artificial (and possibly confusing?) `gh:split_node` tag, and so we do not need the remove operation for the tags (in case we like to store the node tags with EdgeKVStorage or similar).

Now we use the following data structures for the node data:

* idsByOSMNodeIds is a (very large) map that maps the full range of possible (64bit) OSM node IDs to a smaller (32bit) range of indices of nodes we actually use to build the graph (typically those contained in ways using the `highway` tag).

* nodeTagIndicesByOSMNodeIds is a similar, but smaller, map that does the same, but only for nodes for which we keep the tags for

* nodeTags is a list that maps the int indices we get via nodeTagIndicesByOSMNodeIds to the actual tags

* nodesToBeSplit is a separate hash set of OSM node IDS that shall be split. for this we need a way to remove/unset values.

Quite possibly we could combine some of these to either save memory or speed up the import, or both? For example we could combine idsByOsmNodeIds and nodeTagIndicesByOsmNodeIds, which would definitely be an improvement if we stored tags for **all* nodes. Or we could use the int index we get via nodeTagIndicesByOsmNodeIds to replace the nodesToBeSplit LongSet with a BitSet (using the int index).

Reply:
> Quite possibly we could combine some of these to either save memory or speed up the import, or both?

It looks like whenever we do a lookup into nodeTagIndicesByOSMNodeIds we also do a lookup into idsByOSMNodeIds anyway, which means we could get rid of the lookup into nodeTagIndicesByOSMNodeIds. But since there are less nodes for which we store tags than there are nodes we need to map the (larger) int index we get from idsByOSMNodeIds to a smaller, more compact index. At least as long as we want to keep the node tags list compact. So actually we would then need another int-int lookup, and I'm not sure if overall this will be better, but it might be worth a try.

Reply:
Did you check for how many nodes we now store the tag (compared to the total number of nodes we map the coordinates for)?

Reply:
A bit off-topic, but it also might speed up the import if we replaced:

```
 int curr = idsByOsmNodeIds.get(osmNodeId);
 if (curr == EMPTY_NODE)
       idsByOsmNodeIds.put(osmNodeId, newNodeType);
```

with something like
```
  int index = idsByOsmNodeIds.indexOf(osmNodeId);
  int curr = idsByOsmNodeIds.indexGet(index);
  if (curr == EMPTY_NODE)
        idsByOsmNodeIds.indexPut(index, newNodeType);
```

but not sure if using an index like this is possible with GHLongIntBTreeMap.